common/src/ess.rs

```diff
@@ -0,0 +1,190 @@
+            )
+            .map_err(|_| Status::failed_precondition("The specified client does not exist."))?;
+
+        for subscription in subscriptions {
+            let source = subscription.event_id().to_string();
+
+            spawn(subscription.serve(move |data, seq| {
+                Ok(Event {
+                    source: source.clone(),
+                    value: Some(ValueMessage { value: Some(into_value(data)) }),
+                    seq,
+                    timestamp: Some(SystemTime::now().into()),
+                })
+            }));
+        }
+
+        Ok(SubscribeFulfillment {})
+    }
+}
+
+#[async_trait]
+impl<T> ChannelService for SharedEss<T>
+where
+    T: Clone + Send + Sync + 'static,
+{
+    type OpenStream = ReceiverStream<Result<Event, Status>>;
+
+    async fn open(
+        &self,
+        _: tonic::Request<OpenRequest>,
+    ) -> Result<Response<Self::OpenStream>, Status> {
+        const METADATA_KEY: &str = "x-chariott-channel-id";
+
+        let id = Uuid::new_v4().to_string();
+        let (_, receiver_stream) = self.0.read_events(id.clone().into());
+        let mut response = Response::new(receiver_stream);
+        response.metadata_mut().insert(METADATA_KEY, id.try_into().unwrap());
+        Ok(response)
+    }
+}
+
+impl<T> Deref for SharedEss<T> {
+    type Target = InnerEss<T>;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.as_ref()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use crate::proto::{
+        common::Value as ValueMessage,
+        common::{value::Value as ValueEnum, SubscribeIntent},
+        streaming::{channel_service_server::ChannelService, OpenRequest},
+    };
+    use tokio_stream::StreamExt as _;
+    use tonic::{Code, Request};
+
+    use super::SharedEss;
+
+    #[tokio::test]
+    async fn open_should_set_channel_id() {
+        // arrange
+        let subject = setup();
+
+        // act
+        let response = subject.open(Request::new(OpenRequest {})).await.unwrap();
+
+        // assert
+        assert!(!response.metadata().get("x-chariott-channel-id").unwrap().is_empty());
+    }
+
+    #[tokio::test]
+    async fn serve_subscriptions_should_serve_subscription_for_event() {
+        // arrange
+        const EVENT_A: &str = "test-event-a";
+        const EVENT_B: &str = "test-event-b";
+
+        let subject = setup();
+        let response = subject.open(Request::new(OpenRequest {})).await.unwrap();
+        let channel_id =
+            response.metadata().get("x-chariott-channel-id").unwrap().to_str().unwrap().into();
+
+        // act
+        subject
+            .serve_subscriptions(
+                SubscribeIntent { channel_id, sources: vec![EVENT_A.into(), EVENT_B.into()] },
+                |_| ValueEnum::Null(0),
+            )
+            .unwrap();
+
+        // assert
+        subject.0.publish(EVENT_A, ());
+        subject.0.publish(EVENT_B, ());
```

Comment:
nit:

```suggestion
        subject.publish(EVENT_A, ());
        subject.publish(EVENT_B, ());
```

Reply:
Don't these belong in the act part?

Reply:
I don't think so, as these are events only published for assertions. That was the only way I saw to check whether we actually served the correct subscription. Applied your suggesion.